Philips.CodeAnalysis.MaintainabilityAnalyzers/RuntimeFailure/ThrowInnerExceptionAnalyzer.cs

```diff
@@ -0,0 +1,97 @@
+﻿// © 2020 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.RuntimeFailure
+{
+	/// <summary>
+	/// Include the original exception when rethrowing.
+	/// </summary>
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class ThrowInnerExceptionAnalyzer : DiagnosticAnalyzer
+	{
+		private const string Title = "User inner exceptions for unhandled exceptions";
+		private const string Message = "Rethrown exception should include caught exception.";
+		private const string Description = "User inner exceptions for unhandled exceptions";
+		private const string Category = Categories.RuntimeFailure;
+
+		private static readonly DiagnosticDescriptor Rule =
+			new DiagnosticDescriptor(
+				Helper.ToDiagnosticId(DiagnosticIds.ThrowInnerException),
+				Title,
+				Message,
+				Category,
+				DiagnosticSeverity.Error,
+				isEnabledByDefault: true,
+				description: Description
+			);
+
+		/// <summary>
+		/// <inheritdoc/>
+		/// </summary>
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		/// <summary>
+		/// <inheritdoc/>
+		/// </summary>
+		public override void Initialize(AnalysisContext context)
+		{
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.EnableConcurrentExecution();
+			context.RegisterSyntaxNodeAction(
+				AnalyzeCatchException,
+				SyntaxKind.CatchClause
+			);
+		}
+
+		private void AnalyzeCatchException(SyntaxNodeAnalysisContext context)
+		{
+			var catchNode = (CatchClauseSyntax)context.Node;
+			// Look for throw statements and check them.
+			var hasBadThrowNodes = catchNode.DescendantNodes().OfType<ThrowStatementSyntax>()
+				.Where(node => !IsCorrectThrow(context, node)).Any();
+			if (hasBadThrowNodes)
+			{
+				var location = catchNode.GetLocation();
+				var lineNum = Helper.GetLineNumber(location);
+				context.ReportDiagnostic(Diagnostic.Create(Rule, location, lineNum));
+			}
+		}
+
+		// Throw should rethrow same exception, or include original exception
+		// when creating new Exception.
+		private bool IsCorrectThrow(SyntaxNodeAnalysisContext context, ThrowStatementSyntax node)
+		{
+			bool isOk = true;
+			var newNodes = node.ChildNodes().OfType<ObjectCreationExpressionSyntax>();
+			if (newNodes.Any())
+			{
+				// Constructor needs to have at least two arguments.
+				var objectCreation = newNodes.First();
```

Comment:
Is there a reason not to check all of the object creation nodes?  There is nothing (other than common sense) stopping someone from writing

```
throw (DateTime.Now.DayOfWeek == DayOfWeek.Monday ? new Exception("") : new InvalidCastException(""));
```

Reply:
No, checking all of them now